iot-management-device/src/main/kotlin/com/example/iotmanagementdevice/repository/RedisDeviceRepository.kt

```diff
@@ -0,0 +1,129 @@
+                        }
+                }
+            }
+            .switchIfEmpty { findInMongoAndWriteToRedis(deviceId) }
+            .onErrorResume(::isErrorFromRedis) {
+                mongoDeviceRepository.findById(deviceId)
+            }
+    }
+
+    override fun deleteById(deviceId: String): Mono<Unit> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.deleteById(deviceId)
+            .flatMap {
+                reactiveRedisTemplate.unlink(key)
+                    .onErrorResume(::isErrorFromRedis) {
+                        log.warn("Redis failed to remove user from cache", it)
+                        Mono.empty()
+                    }.thenReturn(Unit)
+            }.thenReturn(Unit)
+    }
+
+    private fun saveDeviceToRedis(device: MongoDevice): Mono<MongoDevice> {
+        val key = createDeviceKey(device.id.toString())
+        val byteArray = mapper.writeValueAsBytes(device)
+        return reactiveRedisTemplate.opsForValue().set(
+            key,
+            byteArray,
+            Duration.ofMinutes(redisExpirationTimeoutInMinutes)
+        ).thenReturn(device)
+    }
+
+    private fun saveDeviceWithRetries(device: MongoDevice) {
+        saveDeviceToRedis(device).retryWhen(
+            Retry.backoff(MAX_RETRIES, Duration.ofSeconds(RETRY_BACKOFF_SECONDS))
+                .filter { error -> isErrorFromRedis(error) }
+                .doBeforeRetry { retrySignal ->
+                    log.warn(
+                        "Retrying saving device to Redis because of {}:",
+                        retrySignal.failure().message
+                    )
+                }
+        )
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe()
+    }
+
+    private fun findInMongoAndWriteToRedis(deviceId: String): Mono<MongoDevice> {
+        val key = createDeviceKey(deviceId)
+        return mongoDeviceRepository.findById(deviceId).flatMap {
+            saveDeviceToRedis(it)
+        }
+            .switchIfEmpty {
+                reactiveRedisTemplate.opsForValue()
+                    .set(key, byteArrayOf(), Duration.ofMinutes(redisExpirationTimeoutInMinutes))
+                    .then(Mono.empty())
+            }
+    }
+
+    private fun isErrorFromRedis(throwable: Throwable): Boolean {
+        return redisErrors.any { it.isInstance(throwable) }
+    }
+
+    companion object {
+        private const val KEY_PREFIX = "device"
+        private val log = LoggerFactory.getLogger(RedisDeviceRepository::class.java)
+        private const val MAX_RETRIES = 5L
+        private const val RETRY_BACKOFF_SECONDS = 2L
+
+        private val redisErrors = setOf(
+            RedisConnectionFailureException::class,
+            RedisException::class,
+            SocketException::class,
+            QueryTimeoutException::class,
+        )
+
+        fun createDeviceKey(key: String): String {
+            return "$KEY_PREFIX$key"
```

Comment:
```suggestion
            return "$KEY_PREFIX::$key"
```
it's better to separate prefix and id